From 2c7f2dcdbf7f53078985ff1a997e8c0e6c16fa4f Mon Sep 17 00:00:00 2001
From: Jonas Karlman <jonas@kwiboo.se>
Date: Sun, 27 Apr 2025 20:34:24 +0000
Subject: [PATCH 186/213] RFC: drm/panthor: Add support for a bus clk

On RK3588 the GPU clk is exposed using a normal CLK_GPU and also as
SCMI_CLK_GPU. The main clock for the GPU is the SCMI_CLK_GPU, however,
the normal CLK_GPU also need to be enabled independently.

Add support for a bus clk to handle these two different clocks, similar
to panfrost.

Signed-off-by: Jonas Karlman <jonas@kwiboo.se>
---
 drivers/gpu/drm/panthor/panthor_device.c | 16 +++++++++++++++-
 drivers/gpu/drm/panthor/panthor_device.h |  3 +++
 2 files changed, 18 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/panthor/panthor_device.c b/drivers/gpu/drm/panthor/panthor_device.c
index 9687c59de350..39ab4f8eb5b7 100644
--- a/drivers/gpu/drm/panthor/panthor_device.c
+++ b/drivers/gpu/drm/panthor/panthor_device.c
@@ -47,6 +47,12 @@ static int panthor_clk_init(struct panthor_device *ptdev)
 				     PTR_ERR(ptdev->clks.coregroup),
 				     "get 'coregroup' clock failed");
 
+	ptdev->clks.bus = devm_clk_get_optional(ptdev->base.dev, "bus");
+	if (IS_ERR(ptdev->clks.bus))
+		return dev_err_probe(ptdev->base.dev,
+				     PTR_ERR(ptdev->clks.bus),
+				     "get 'bus' clock failed");
+
 	drm_info(&ptdev->base, "clock rate = %lu\n", clk_get_rate(ptdev->clks.core));
 	return 0;
 }
@@ -499,10 +505,14 @@ int panthor_device_resume(struct device *dev)
 
 	atomic_set(&ptdev->pm.state, PANTHOR_DEVICE_PM_STATE_RESUMING);
 
-	ret = clk_prepare_enable(ptdev->clks.core);
+	ret = clk_prepare_enable(ptdev->clks.bus);
 	if (ret)
 		goto err_set_suspended;
 
+	ret = clk_prepare_enable(ptdev->clks.core);
+	if (ret)
+		goto err_disable_bus_clk;
+
 	ret = clk_prepare_enable(ptdev->clks.stacks);
 	if (ret)
 		goto err_disable_core_clk;
@@ -561,6 +571,9 @@ int panthor_device_resume(struct device *dev)
 err_disable_core_clk:
 	clk_disable_unprepare(ptdev->clks.core);
 
+err_disable_bus_clk:
+	clk_disable_unprepare(ptdev->clks.bus);
+
 err_set_suspended:
 	atomic_set(&ptdev->pm.state, PANTHOR_DEVICE_PM_STATE_SUSPENDED);
 	atomic_set(&ptdev->pm.recovery_needed, 1);
@@ -608,6 +621,7 @@ int panthor_device_suspend(struct device *dev)
 	clk_disable_unprepare(ptdev->clks.coregroup);
 	clk_disable_unprepare(ptdev->clks.stacks);
 	clk_disable_unprepare(ptdev->clks.core);
+	clk_disable_unprepare(ptdev->clks.bus);
 	atomic_set(&ptdev->pm.state, PANTHOR_DEVICE_PM_STATE_SUSPENDED);
 	return 0;
 }
diff --git a/drivers/gpu/drm/panthor/panthor_device.h b/drivers/gpu/drm/panthor/panthor_device.h
index 0fda64fbe5f2..7f41aa71e287 100644
--- a/drivers/gpu/drm/panthor/panthor_device.h
+++ b/drivers/gpu/drm/panthor/panthor_device.h
@@ -142,6 +142,9 @@ struct panthor_device {
 
 	/** @clks: GPU clocks. */
 	struct {
+		/** @bus: Bus clock. This clock is optional. */
+		struct clk *bus;
+
 		/** @core: Core clock. */
 		struct clk *core;
 
-- 
2.43.0

